sqldelight-gradle-plugin/src/test/kotlin/com/squareup/sqldelight/dialect/DialectIntegrationTests.kt

```diff
@@ -12,6 +12,7 @@ class DialectIntegrationTests {
   @Test fun integrationTestsMySql() {
     val runner = GradleRunner.create()
       .withCommonConfiguration(File("src/test/integration-mysql"))
+      .withPluginClasspath()
```

Comment:
I'm not a fan of the use of this mechanism. It is _very_ broken. We started out using it years ago and migrated away because of its problems. The classpath of this plugin is injected as a set of resolved artifacts which do not participate in the normal buildscript classpath resolution semantics of the project under test.

The two problems I remember were:
- It includes compileOnly dependencies in this classpath making them not actually behave like compileOnly.
- It means multiple copies of the same dependency can make their way onto the buildscript classpath since one will come from the file generated by this mechanism and the other will come in through the normal dependency mechanism.

Reply:
I'm not sure the first problem is accurate, [this](https://docs.gradle.org/7.2/javadoc/org/gradle/plugin/devel/tasks/PluginUnderTestMetadata.html#getPluginClasspath--) says the injected classpath defaults to `sourceSets.main.runtimeClasspath` which is what you'd want. A cursory look over the generated `plugin-under-test-metadata.properties` doesn't show any compileOnly deps.

On the second, I understand this could happen if there's a mix of `plugins {}` and `buildscript.dependencies { classpath ... }` used in the test project build files, but it shouldn't be an issue if everything in `plugins {}` is injected and the buildscript mechanism isn't used. When it's setup this way everything is injected and nothing is resolved externally (unless one of the injected plugins resolves other build dependencies... SqlDelight doesn't seem to but I don't know about AGP).

If it's a hard no then I'll just revert the first commit, or at least what's changed in the plugin's /src/test

Reply:
Ah you're right about the first problem, I'm misremembering. We used to not use `compileOnly` and had a hard `api` dependency on the Android plugin. We also would add the Android plugin in these fixture tests (granted, using the buildscript classpath and not the plugins mechanism) which would pull from an external repository. Inside the Android Gradle plugin there is code which ensures that its dependencies are coming from the same classloader which was failing because we were causing two copies to be on the classpath (one from normal buildscript classpath dependencies and one injected by the plugin classpath text file).

In general, I prefer to test projects as close as possible to how they will actually be used. I am okay with migrating fixtures to the plugins block. I am still wary of the use of this mechanism vs. installing to our own Maven repository (as we already must do for the runtime artifacts).

Reply:
> I prefer to test projects as close as possible to how they will actually be used

Fair.

I've used TestKit & its classpath injection in anger on another project and it has some issues (like the fact that the runtime artifacts can't be injected, and multiple classloaders [which also bit when some plugins are injected by TestKit and depend on other plugins applied in `plugins {}` that aren't injected - they apparently have isolated classloaders which is very different to a real build) but otherwise it's been OK. This project is much more complex though, and if it ain't broke...

Instead, what do you think about adding `settings.gradle.kts` with:
```kotlin
pluginManagement {
    repositories {
        maven(url = "file://path/to/localMaven")
        gradlePluginPortal()
    }
}
```
Then the plugins are applied with `plugins {}`, resolve to local Maven, and the test `buildscript.gradle` is only needed for Android tests. That way, plugins are applied using the standard plugin mechanism. This is actually probably closest to a real build config.

Reply:
i will throw one additional hit into the ring of not using `withPluginClasspath()` - I think it means we're not testing the shaded version of the gradle plugin on CI, which means its tested nowhere which I do not have confidence in staying stable

Reply:
im definitely okay with your `settings.gradle.kts` idea